File: questions/q134_maximum_stream_subarray/deque.cpp
// This is for the leetcode version of the problem


class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        deque<int> store;
        vector<int> ans;

        for (int i=0; i<nums.size(); i++) {
            while (store.size() > 0 && nums[store.back()] <= nums[i]) {
                store.pop_back();
            }
            while (store.size() > 0 && store.front() <= i-k) {
                store.pop_front();
            }
            store.push_back(i);
            
            if (i >= k-1) {
                ans.push_back(nums[store.front()]);
            }
        }

        return ans;
    }
};

File: questions/q134_maximum_stream_subarray/deque.py
# This is for the leetcode version of the problem


from collections import deque

class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        store = deque()
        ans = []

        for i in range(len(nums)) :
            while len(store) > 0 and nums[store[-1]] <= nums[i] :
                store.pop()
            while len(store) > 0 and store[0] <= i-k :
                store.popleft()
            store.append(i)

            if i >= k-1 :
                ans.append(nums[store[0]])

        return ans
